### PR TITLE
view_channel_3d: take an already created figure as input

### PR DESCRIPTION
**Pull request by Edouard2laire**
This PR allows the following code to work: 

```matlab

hFig  = bst_figures('GetFigureWithSurface', SurfaceFile);
hFig  = view_surface_data(SurfaceFile, OverlayFile, [], hFig, 1);
hFig  = view_surface(ScalpFile, 0, [], hFig, 1);
hFig  = view_channels_3d(channelFile, 'NIRS-BRS', 'scalp', 0, 1,hFig);

```

This code produce the following image and solve my problem from here: https://neuroimage.usc.edu/forums/t/visualizing-sensors-on-top-of-cortical-activity/44950/2

image

hs: is there a better way to create an empty figure than : hFig bst_figures('GetFigureWithSurface', SurfaceFile); ?

### CODE REVIEW

**Comment by rcassani** [comment]
> hs: is there a better way to create an empty figure than : hFig = bst_figures('GetFigureWithSurface', SurfaceFile); ?

Not sure what is the question here. If the surface is not plotted in any figure, hFig is just empty, no figure is created. Otherwise, hFig is the handle of the figure where the surface is plotted.

**Comment by rcassani** [comment]
> `'NewFigure'`

That's why I got confused, that option is just there to be used

**Comment by rcassani** [comment]
@Edouard2laire with the latest changes, the image can be plot as:

```
hFig  = view_surface_data(SurfaceFile, OverlayFile, [], 'NewFigure', 1);
hFig  = view_channels_3d(channelFile, 'NIRS-BRS', 'scalp', 0, 1, hFig);
```

And if the scalp is not needed, just leave that string empty `''`
I think this is ready to be merge, let me know if you agree

**Comment by Edouard2laire** [comment]
> > `'NewFigure'`
> 
> That's why I got confused, that option is just there to be used


Yes, I don't know why didn't see that at first and was trying very complex things... haha

Yes, it's working. 
Thanks a lot

**Comment by chinmaychinara91** [comment]
@rcassani @Edouard2laire this change is creating new figures all the time. I am actually using the same function for SEEG work.

In `tree_callbacks.m`, when I modify this:

```
elseif strcmpi(DisplayMod{1}, 'SEEG')
    hFig = bst_figures('GetFiguresByType', {'3DViz'});
    hFig = DisplayChannels(bstNodes, DisplayMod{1}, 'anatomy', 1, 0, hFig);
```
The `DisplayChannels` function uses `view_channels_3d.m`

So if I load a channel file and double click on the channel file shown below, rather than just updating the figure already displayed it keep generating new figures. Let me know if this is expected.
Screenshot 2024-02-29 132959

**Comment by Edouard2laire** [comment]
> @rcassani @Edouard2laire this change is creating new figures all the time. I am actually using the same function for some SEEG work.
> 
> In `tree_callbacks.m`, when I modify this:
> 
> ```
> elseif strcmpi(DisplayMod{1}, 'SEEG')
>     hFig = bst_figures('GetFiguresByType', {'3DViz'});
>     hFig = DisplayChannels(bstNodes, DisplayMod{1}, 'anatomy', 1, 0, hFig);
> ```
> 
> The `DisplayChannels` function uses `view_channels_3d.m`
> 


Oh that is interesting.  I am not sure if that is expected.  Can you check if hfig is sent to view_channels_3d when called through  DisplayChannels ?

**Comment by chinmaychinara91** [comment]
> > @rcassani @Edouard2laire this change is creating new figures all the time. I am actually using the same function for some SEEG work.
> > In `tree_callbacks.m`, when I modify this:
> > ```
> > elseif strcmpi(DisplayMod{1}, 'SEEG')
> >     hFig = bst_figures('GetFiguresByType', {'3DViz'});
> >     hFig = DisplayChannels(bstNodes, DisplayMod{1}, 'anatomy', 1, 0, hFig);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The `DisplayChannels` function uses `view_channels_3d.m`
> 
> Oh that is interesting. I am not sure if that is expected. Can you check if hfig is sent to view_channels_3d when called through DisplayChannels ?

I did check it is sent. I think I found the issue:
in `view_channels_3d` if we do a `if strcmpi(hFig, 'NewFigure')` check before every:
`hFig = view_mri_3d(SurfaceFile, [], SurfAlpha, hFig);` that kind of solves it

**Comment by rcassani** [comment]
> @rcassani @Edouard2laire this change is creating new figures all the time. I am actually using the same function for some SEEG work.

I may be wrong, but isn't this the current behaviour in the master branch?

**Comment by chinmaychinara91** [comment]
> > @rcassani @Edouard2laire this change is creating new figures all the time. I am actually using the same function for some SEEG work.
> 
> I may be wrong, but isn't this the current behaviour in the master branch?

@rcassani let me know if the suggestion int the previous comment works

**Comment by Edouard2laire** [comment]
I am curious as to why the if is needed. If hFig is a valid figure, then hFig = view_mri_3d(SurfaceFile, [], SurfAlpha, hFig); should not create a new figure; maybe its an issue to fix inside view_mri_3d instead of adding an if.

**Comment by rcassani** [comment]
@chinmaychinara91, the double plotting comes from `view_mri_3d` at [L133-L137](https://github.com/brainstorm-tools/brainstorm3/blob/master/toolbox/gui/view_mri_3d.m#L133-L137)

The proposed solution kind of works, as skips to plot the 'anatomy' if a handle is given, but it does not check if the anatomy (3D slices) are already there. It will not work properly with:

```
hFig  = view_surface(ScalpFile, 0, [], [], 1);
hFig  = view_channels_3d(channelFile, 'SEEG', 'anatomy', 0, 1, hFig);
```

How do like the proposed change in [916bc36](https://github.com/brainstorm-tools/brainstorm3/pull/688/commits/916bc36cad67da1f43a314fde2166f9958b5fd34)?

---
> If hFig is a valid figure, then hFig = view_mri_3d(SurfaceFile, [], SurfAlpha, hFig); should not create a new figure; maybe its an issue to fix inside view_mri_3d instead of adding an if. 

yes, it seems that was the case

File: toolbox/gui/view_channels_3d.m [hunk 1]
```diff
@@ -136,20 +132,17 @@
...
                 if isempty(SurfaceFile)
                     SurfaceFile = sSubject.Anatomy(sSubject.iAnatomy).FileName;
                 end
                 SurfAlpha = .1;
-                if isempty(hFig)
-                    hFig = view_mri_3d(SurfaceFile, [], SurfAlpha, 'NewFigure');
-                end
+                hFig = view_mri_3d(SurfaceFile, [], SurfAlpha, hFig);
             end
+        otherwise
```
Comment (Edouard2laire) [inline toolbox/gui/view_channels_3d.m:145]: Should an error be thrown in this case?

Comment (rcassani) [reply toolbox/gui/view_channels_3d.m:145]: I left it like this with the idea that any invalid surface type will just plot the sensors

**Comment by chinmaychinara91** [comment]
> @chinmaychinara91, the double plotting comes from `view_mri_3d` at [L133-L137](https://github.com/brainstorm-tools/brainstorm3/blob/master/toolbox/gui/view_mri_3d.m#L133-L137)
> 
> The proposed solution kind of works, as skips to plot the 'anatomy' if a handle is given, but it does not check if the anatomy (3D slices) are already there. It will not work properly with:
> 
> ```
> hFig  = view_surface(ScalpFile, 0, [], [], 1);
> hFig  = view_channels_3d(channelFile, 'SEEG', 'anatomy', 0, 1, hFig);
> ```
> 
> How do like the proposed change in [916bc36](https://github.com/brainstorm-tools/brainstorm3/pull/688/commits/916bc36cad67da1f43a314fde2166f9958b5fd34)?
> 
> > If hFig is a valid figure, then hFig = view_mri_3d(SurfaceFile, [], SurfAlpha, hFig); should not create a new figure; maybe its an issue to fix inside view_mri_3d instead of adding an if.
> 
> yes, it seems that was the case

works perfect. Thanks @rcassani.

**Comment by rcassani** [comment]
@chinmaychinara91, @Edouard2laire 
This is a nice new feature, thank you both!

**Pull request merged**